### PR TITLE
Split rerouting from TransportReplicationAction

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today's `TransportReplicationAction` routes the request to the primary
(following relocations etc) and then performs the action and replicates it. In
some cases we want to avoid the rerouting phases because the action is being
coordinated by the primary itself. This change separates out a
`TransportReroutedReplicationAction` which performs the rerouting and primary
delegation phases before handing over to `TransportReplicationAction` to do the
actual replication.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Note that this doesn't actually use `TransportReplicationAction` in isolation anywhere yet. It's already quite a substantial change without that, and I'd like some feedback before I commit to things any further.

**Comment by DaveCTurner** [comment]
Failure looks like we were very slow to start up the nodes for an integ test (>4 minutes). @elasticmachine please run elasticsearch-ci/1.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 1]
```diff
@@ -16,182 +16,109 @@
...
-
-    /**
-     * Primary operation on node with primary copy.
-     *
-     * @param shardRequest the request to the primary shard
-     * @param primary      the primary shard to perform the operation on
-     */
-    protected abstract void shardOperationOnPrimary(Request shardRequest, IndexShard primary,
-        ActionListener<PrimaryResult<ReplicaRequest, Response>> listener);
+    protected abstract void acquireReplicaOperationPermit(IndexShard replica,
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: Perhaps we should just add a boolean to this method to denote whether one or all permits should be acquired, we can then avoid exposing this method to subclasses and the extra delegation required here.

**Review by ywelsch — COMMENTED** [review]
The split makes sense to me, but there's quite a bit of ceremony still involved in passing things from `TransportReroutedReplicationAction` to `TransportReplicationAction` in order to allow extension and adaptation of behavior. I would like to see more proof that the changes here are a good fit for the consumers.

Can you check if `TransportWriteAction` can be changed so that it subclasses `TransportReplicationAction` (the write action  is all about adapting the behavior of the replication bit, not the reroute bit).
On a similar note, `TransportResyncReplicationAction` is something that does not need rerouting. Can it be adapted to purely extend TransportReplicationAction (or the new `TransportWriteAction`)?

**Comment by DaveCTurner** [comment]
Closing this for now until we've time to work on it again.

**Pull request closed without merging**